src/mo_frontend/typing.ml

```diff
@@ -135,6 +135,14 @@ let error_in modes env at code fmt =
 
 let plural cs = if T.ConSet.cardinal cs = 1 then "" else "s"
 
+let warn_lossy_bind_type env at bind t1 t2 =
+  let common = T.lub t1 t2 in
+  if T.(not (eq t1 common || eq common t2)) then
```

Comment:
I think else where Andreas used 'inconsistent common [t1; t2]` which I think only warns if the lub degrades to `Any` but none of the branches have type `Any`.

Reply:
Right, but that is too coarse for my taste.

Reply:
Right. That would warn for `(x : Nat) or (x : Bool)` (which is inconsistent anyway) but not for `(x : (Int, Nat)) or (x : (Nat, Int))`. In a language with subtyping, it's always a bit fishy to make subtyping an error arbitrarily, though for inferred bindings the situation arguably is different.

As an aside, you can simplify your condition to
```suggestion
  if not T.(sub t1 t2 || sub t2 t1) then
```
which should be slightly more efficient.

Reply:
Thanks, I missed this comment before merging, but it is applied now on the other branch as 1655c8ef1.

Reply:
Thinking about this some more, I'd suggest to change the logic to use `inconsistent` after all. It is weird to accept
```
(x1 : Nat, x2 : Int) or (x1 : Int, x2 : Nat)`
```
without warning, but warn about
```
(x : (Nat, Int)) or (x : (Int, Nat))`
```
That is the kind of thing that violates compositionality, and it is annoying when seemingly innocent refactorings of code can introduce warnings.